### ISSUE TITLE
Preferences are not saved

### ISSUE THREAD

**Issue by bsiegel** [JohnCoates/Aerial#917]
#### General troubleshooting tips

Before logging an issue please check that:
- [x] You have the latest version installed (There may be a beta version that fixes your issue), see here for the latest releases and bug fixes: https://github.com/JohnCoates/Aerial/releases
- [x] Your issue isn't already mentioned in our [issues](https://github.com/JohnCoates/Aerial/issues). You may find a workaround there or a similar request already made.
- [x] Your problem isn't mentioned in the [troubleshooting section](https://github.com/JohnCoates/Aerial#troubleshooting).

If none of this fixes your issue, tell us about the problem you are experiencing or the feature you'd like to request.

#### Required information
In order to help us sort your issue, we ask that you provide the following information:
- [x] Mac model: MacBook Pro 15-inch, 2017
- [x] macOS version: 10.15.2
- [x] Monitor setup: 2x external DisplayPort displays

If appropriate, please enable `Debug mode` and `Log to disk` in `Advanced` tab and replicate your bug, then attach the `AerialLog.txt` file (You can access this file through the Advanced tab).

[AerialLog.txt](https://github.com/JohnCoates/Aerial/files/4161670/AerialLog.txt)

`com.JohnCoates.Aerial.plist`:
```xml
 <?xml version="1.0" encoding="UTF-8"?>
 <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
 <plist version="1.0">
 <dict>
     <key>SUEnableAutomaticChecks</key>
     <true/>
     <key>SUHasLaunchedBefore</key>
     <true/>
     <key>SULastCheckTime</key>
     <date>2020-02-05T04:52:46Z</date>
 </dict>
 </plist>
```

#### Description of issue

I've tested on 1.6.4, 1.7.0, and 1.7.1 beta. On all versions, no changes that I make in the Screen Saver Options are saved upon clicking "Done". If I come back to the options immediately I see the changes, but if I close System Preferences and re-launch it, all the settings are reset to their defaults. I also don't see them being saved to the preferences plist. I don't see anything in the logs that would explain what is happening.

I will test with some older versions to see if I can locate at what point this behavior began. I was a long time user of the old version (1.2) and only recently came by and noticed that there had been much activity since that version was released. I installed 1.7.0 and deleted the entire `legacyScreenSaver` container just to make sure that nothing was left over.

**Comment by glouel** [comment]
Hi @bsiegel 

This is super weird, but a few thoughts : 
- First, in Catalina, the location of screensavers settings have moved to the container, so your old settings are ignored or were deleted. 
- Second, there are actually two separate plist. The one you mention is the one used by Sparkle (the auto updater that doesn't work in Catalina...). 

The "correct" plist is here : `~/Library/Containers/com.apple.ScreenSaver.Engine.legacyScreenSaver/Data/Library/Preferences/ByHost/com.JohnCoates.Aerial.{{UUID}}.plist`

This is the one that actually stores your config. Can you check that one and get back to me ? 
- As for changes "remaining" when clicking done and coming back, at that point the screensaver is still in memory so that makes sense (it doesn't reload the prefs).

It's still super weird and I haven't had a single report like this so this seems to be something very specific.

I checked your log but nothing specific popped to my mind. Could you please check the plist I mentioned above, does it exist at all, does the content inside change or not (you definitely need to press "Done" to save settings by the way, but you mentionned you did), etc. 

At that point I'm not super sure what's happening, so any extra detail is welcome. And I wouldn't recommend the backtracking of versions, this is probably something more weird considering nobody hit the issue thus far (it could be because you jumped from 1.2 to a modern version although I have a hard time understanding why, etc). 

Sorry about that in any case.

**Comment by bsiegel** [comment]
Just double checked, there is no `ByHost` folder within `Preferences` at all. I tried widening the permissions on the `Preferences` directory in the container (it was 700, I changed it to 755) and relaunched System Preferences -> Screen Saver Options, changed some settings, and clicked Done. But it still did not result in a ByHost directory or a preferences file being created.

**Comment by glouel** [comment]
That could explain it, but it's super weird. As far as I know, it's not Aerial's job to create the ByHost folder (maybe it is and I'm mistaken, but the APIs don't work like this from my possibly limited understanding).

May I suggest you create the "ByHost" folder ? Here's how it shows up here : 

```
drwx------   4 glouel  staff   128 30 jan 00:14 ByHost
```

I think that should solve it, although this is completely puzzling to me how that happened to you... 

I have to ask if you run SIP disabled by the way, or some other form of "enterprise" tool ?

Again my apologies, but that's a first and I'm quite puzzled by that...

**Comment by bsiegel** [comment]
Thanks @glouel, that did the trick. Manually creating the ByHost directory and then relaunching / saving the Screen Saver Options resulted in a plist file being created within the ByHost directory, and the preferences appear to be saved now across re-launches of System Preferences.

I have no idea why that folder was not being created by the OS when the container was first created. I haven't made any unusual changes to my system - SIP is enabled. The device is enrolled with MDM (Intune) which applies some configuration profiles. However nothing there looks suspicious, the only one that makes any setting changes is a Password Policy to enforce password complexity requirements.

**Comment by glouel** [comment]
Good to hear. This is still a bit puzzling to me. I did install Aerial on a fresh Catalina install not long ago and it definitely worked as intended, creating that folder. So I'm wondering if it's a weird edge case/byproduct of your MDM settings or something like that ? I didn't ask you to check the Console.app's logs for anything weird (it's so full of stuff in Catalina, pretty hard to parse) but that could maybe have given us a clue.

I'll try to get a fresh install in a VM and see if I can reproduce, maybe it's just a random Catalina bug, but I tend to get reports about things like that when they break at large scale, hmm.

Anyway, glad it got fixed, I'll keep an eye to see if I can reproduce it.

**Comment by PatTheMav** [comment]
Can confirm to have the same issue and creating the `ByHost`folder fixed it. I usually install custom software via HomeBrew's casks, so maybe that might contribute to this issue.

**Comment by glouel** [comment]
Ha, this may be it @PatTheMav ! It's quite possible that by copying the screensaver instead of double clicking it, whatever machinery they put in place that creates the sandbox omits the ByHost folder ? 

I'll put a mention about it somewhere in the readme, thanks for the heads up !

**Comment by PatTheMav** [comment]
@glouel Yeah I haven't read up on current specs for the Cask format, maybe the Cask needs to be adapted/changed to adhere to Screensavers' sandbox regulations.

**Comment by glouel** [comment]
I did add a couple more mentions just in case (I already had it in troubleshooting but I didn't mention homebrew). 

It's Catalina that creates the sandbox and it's content, so I don't think there's much we can do in the cask. Apple will (hopefully) get rid of that whole mess and allow 3rd party devs to use the new screensaver API in the next macOS release. So many things are broken right now !
